torch/csrc/jit/irparser.cpp

```diff
@@ -0,0 +1,446 @@
+  void parseBlockOutputs(Block* b);
+
+  void parseOperatorsList(Block* b);
+  void parseOperator(Block* b);
+  void parseOperatorOutputs(std::vector<VarWithType>* outs);
+  std::string parseOperatorName();
+  void parseOperatorInputs(Node* n);
+  void parseAttrs(Node* n);
+  void parseAttr(Node* n);
+
+  void parseList(
+      int begin,
+      int sep,
+      int end,
+      const std::function<void()>& callback);
+
+  torch::jit::script::Lexer L;
+  torch::jit::Graph* g = nullptr;
+  std::unordered_map<std::string, Value*> vmap;
+};
+
+struct ParsedLiteral {
+  AttributeKind k = AttributeKind::t;
+
+  long i;
+  std::string s;
+  double f;
+  std::vector<long> is;
+  std::vector<std::string> ss;
+  std::vector<double> fs;
+};
+
+struct VarWithType {
+  VarWithType(){};
+  std::string name;
+  std::string type;
+};
+
+void parseIR(const std::string& str, torch::jit::Graph* graph) {
+  torch::jit::script::IRParser p(str, graph);
+  p.parse();
+}
+
+TypePtr parseType(const std::string& s) {
+  if (s == "Tensor" || s == "NoType") {
+    return TensorType::get();
+  }
+  if (s == "int") {
+    return IntType::get();
+  }
+  if (s == "float") {
+    return FloatType::get();
+  }
+  if (s == "string") {
+    return StringType::get();
+  }
+  // TODO: Support other types.
+  AT_ASSERTM(false, "Type not supported by parser:", s);
+}
+
+VarWithType IRParser::parseVarWithType() {
+  L.expect('%');
+  VarWithType r;
+  if (L.cur().kind == TK_IDENT) {
+    r.name = L.expect(TK_IDENT).text();
+  } else {
+    r.name = L.expect(TK_NUMBER).text();
+  }
+  r.type = "NoType";
+  if (L.nextIf(':')) {
+    r.type = L.expect(TK_IDENT).text();
+  }
+  return r;
+}
+
+std::string IRParser::parseVar() {
+  L.expect('%');
+  if (L.cur().kind == TK_IDENT) {
+    return L.expect(TK_IDENT).text();
+  }
+  return L.expect(TK_NUMBER).text();
+}
+
+void IRParser::parseOperatorOutputs(std::vector<VarWithType>* outs) {
+  if (L.cur().kind != '%') {
+    return;
+  }
+  parseList(TK_NOTHING, ',', TK_NOTHING, [&] {
+    outs->push_back(parseVarWithType());
+  });
+  L.expect('=');
+}
+
+// Parse string or numeric literal and return it along with its type.
+ParsedLiteral IRParser::parseScalarLiteral(Node* n) {
+  auto token = L.cur();
+  std::string str;
+  ParsedLiteral r;
+  switch (token.kind) {
+    case TK_STRINGLITERAL:
+      r.k = AttributeKind::s;
+      r.s = parseStringLiteral(token.range, token.text());
+      L.next();
+      return r;
+    case '-':
+      str = "-";
+      L.next();
+      L.expect(TK_NUMBER);
+      // Fallthrough
+    case TK_NUMBER:
+      str += L.cur().text();
+
+      if (str.find('.') != std::string::npos ||
+          str.find('e') != std::string::npos) {
+        r.k = AttributeKind::f;
+        r.f = std::stod(str);
+      } else {
+        r.k = AttributeKind::i;
+        r.i = std::stoll(str);
+      }
+      L.next();
+      return r;
+    default:
+      std::cout << "Could not parse literal " << token.text() << std::endl;
```

Comment:
All the errors should use `throw ErrorReport(token.range()) << "..."` that way the error is reported with the right source information.